src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb

```diff
@@ -3036,43 +3190,53 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
+            Dim accessor = Me.GetAccessorBlock(declaration, kind)
+            If accessor IsNot Nothing Then
+                accessor = DirectCast(Me.WithStatements(accessor, statements), AccessorBlockSyntax)
+                Return Me.WithAccessorBlock(declaration, kind, accessor)
+            ElseIf Me.CanHaveAccessors(declaration.Kind)
+                accessor = Me.AccessorBlock(kind, statements, Me.GetType(declaration))
+                Return Me.WithAccessorBlock(declaration, kind, accessor)
+            Else
+                Return declaration
+            End If
+        End Function
+
+        Private Function GetAccessorBlock(declaration As SyntaxNode, kind As SyntaxKind) As AccessorBlockSyntax
             Select Case declaration.Kind
                 Case SyntaxKind.PropertyBlock
-                    Dim pb = DirectCast(declaration, PropertyBlockSyntax)
-                    Dim accessor = AccessorBlock(kind, statements, pb.PropertyStatement.AsClause?.Type)
-                    Return pb.WithAccessors(SyntaxFactory.List(WithAccessorBlock(pb.Accessors, accessor)))
-                Case SyntaxKind.PropertyBlock
-                    Dim eb = DirectCast(declaration, EventBlockSyntax)
-                    Dim accessor = AccessorBlock(kind, statements, eb.EventStatement.AsClause?.Type)
-                    Return eb.WithAccessors(SyntaxFactory.List(WithAccessorBlock(eb.Accessors, accessor)))
+                    Return DirectCast(declaration, PropertyBlockSyntax).Accessors.FirstOrDefault(Function(a) a.IsKind(kind))
+                Case SyntaxKind.EventBlock
+                    Return DirectCast(declaration, EventBlockSyntax).Accessors.FirstOrDefault(Function(a) a.IsKind(kind))
                 Case Else
-                    Return declaration
+                    Return Nothing
             End Select
         End Function
 
-        Private Function WithAccessorBlock(accessors As SyntaxList(Of AccessorBlockSyntax), accessor As AccessorBlockSyntax) As SyntaxList(Of AccessorBlockSyntax)
-            Dim currentAccessor = accessors.FirstOrDefault(Function(a) a.IsKind(accessor.Kind))
+        Private Function WithAccessorBlock(declaration As SyntaxNode, kind As SyntaxKind, accessor As AccessorBlockSyntax) As SyntaxNode
+            Dim currentAccessor = Me.GetAccessorBlock(declaration, kind)
             If currentAccessor IsNot Nothing Then
-                Return accessors.Replace(currentAccessor, currentAccessor.WithStatements(accessor.Statements))
-            Else
-                Return accessors.Add(accessor)
+                Return Me.ReplaceNode(declaration, currentAccessor, accessor)
+            ElseIf accessor IsNot Nothing
+                Select Case declaration.Kind
+                    Case SyntaxKind.PropertyBlock
+                        Dim pb = DirectCast(declaration, PropertyBlockSyntax)
+                        Return pb.WithAccessors(pb.Accessors.Add(accessor))
```

Comment:
If WithAccessor is called with a set accessor on a reaonly property, should the readonly modifier be removed from the property or should the caller make sure to do that?